# Project 6: Create the backend of a LLM chatbot

### Backend chatbot

- Dataset link - https://docs.gitbook.com/

### Objective of the project:  Create the backend of a LLM chatbot with Gitbook Documentation as Knowledge Base.

* Use Open Source Large Language Model & Embedding.
* Accomplish the above using Retrieval Augmented Generation Workflow using Langchain Framework.


## Steps


1. Understand the Business Requirements and Data. GitBook is a platform for creating and sharing technical documentation. It provides a collaborative environment for teams to work on documentation projects efficiently.
2. Data Collection.
3. Model selection / Data preparation
* Implement RAG with LangChain
  * Tokenize. Technique to convert text into numbers.
  * Chunking data. Memorization and learning technique to divide information into smaller, more manageable blocks or units, known as 'chunks'.
  * Encoding text
  * Select Data Base
  * Store embeding data (Vector Data Base)
  * Encode similarities
4. Ask the model
  


## I.- Import / Install libraries

In [1]:
!  pip install langchain # A framework to implement RAG


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.6/50.6 kB 1.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 21.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 407.7/407.7 kB 26.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 296.7/296.7 kB 22.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.4/76.4 kB 6.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.0/78.0 kB 5.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 144.5/144.5 kB 12.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.5/54.5 kB 4.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.3/58.3 kB 5.4 MB/s eta 0:00:00


In [2]:
!pip install langchain-community # Framework to implement RAG

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.4/2.4 MB 36.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.5/49.5 kB 3.9 MB/s eta 0:00:00


In [3]:
# Encoding using Transformers (OpenSource Encoding), what is the average token size we get?
# The text we get from the content should be encoded. The entire text should be broken into smaller chunks.
!pip install transformers

In [4]:
!pip install sentence-transformers

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 255.8/255.8 kB 4.6 MB/s eta 0:00:00


In [5]:
#Installing Pinecode Vector Database and dependencies for using it with LangChain
!pip install -qU  pinecone-client==3.0.0  langchain-pinecone==0.0.3

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 199.9/199.9 kB 6.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.9/302.9 kB 16.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 53.0/53.0 kB 4.8 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
langchain 0.3.4 requires langchain-core<0.4.0,>=0.3.12, but you have langchain-core 0.1.52 which is incompatible.
langchain-community 0.3.3 requires langchain-core<0.4.0,>=0.3.12, but you have langchain-core 0.1.52 which is incompatible.
langchain-text-splitters 0.3.0 requires langchain-core<0.4.0,>=0.3.0, but you have langchain-core 0.1.52 which is incompatible.


In [6]:
!pip install auto_gptq

INFO: pip is looking at multiple versions of multiprocess to determine which version is compatible with other requirements. This could take a while.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.5/23.5 MB 81.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 320.7/320.7 kB 25.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 471.6/471.6 kB 33.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.2/13.2 MB 102.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 6.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 8.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 15.9 MB/s eta 0:00:00


In [7]:
!pip install optimum

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 422.6/422.6 kB 2.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.0/46.0 kB 152.6 kB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.8/86.8 kB 7.0 MB/s eta 0:00:00


In [8]:
!pip install -U accelerate bitsandbytes datasets peft transformers

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.4/44.4 kB 2.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 330.9/330.9 kB 13.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 122.4/122.4 MB 5.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.9/9.9 MB 23.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 51.3 MB/s eta 0:00:00
  Attempting uninstall: tokenizers
    Found existing installation: tokenizers 0.19.1
    Uninstalling tokenizers-0.19.1:
      Successfully uninstalled tokenizers-0.19.1
  Attempting uninstall: accelerate
    Found existing installation: accelerate 0.34.2
    Uninstalling accelerate-0.34.2:
      Successfully uninstalled accelerate-0.34.2
  Attempting uninstall: transformers
    Found existing installation: transformers 4.44.2
    Uninstalling transformers-4.44.2:
      Successfully uninstalled transformers-4.44.2


In [9]:
import pandas as pd
from tqdm.auto import tqdm
import hashlib # To generate unique id´s
from uuid import uuid4
from langchain_community.document_loaders import GitbookLoader
from langchain.vectorstores import Pinecone
import nest_asyncio # to allow nested event loops.
from transformers import AutoTokenizer # To control the number of tokens being utilised.
from sentence_transformers import SentenceTransformer
from langchain.text_splitter import RecursiveCharacterTextSplitter #Splits the entire text into smaller portions
from pinecone import Pinecone, ServerlessSpec
import optimum
import auto_gptq
from transformers import AutoModelForCausalLM,AutoTokenizer, pipeline

## II.- Fuctions definition

In [10]:
# Convert text into number
def tokenizer_len(text):
  return len(tokenizer(text, return_tensors = 'pt').input_ids.cuda()[0])

In [11]:
# We need to encode the question that the user asks. So we need to create a function.
def embed_query(text):
  embedding = embedding_model.encode(text).tolist()
  return embedding

## III.- Data collection

In [12]:
nest_asyncio.apply()
loader = GitbookLoader("https://docs.gitbook.com", load_all_paths=True)
all_pages_data = loader.load()

/usr/lib/python3.10/html/parser.py:170: XMLParsedAsHTMLWarning: It looks like you're parsing an XML document using an HTML parser. If this really is an HTML document (maybe it's XHTML?), you can ignore or filter this warning. If it's XML, you should know that using an XML parser will be more reliable. To parse this document as XML, make sure you have the lxml package installed, and pass the keyword argument `features="xml"` into the BeautifulSoup constructor.
  k = self.parse_starttag(i)
Fetching pages: 100%|##########| 108/108 [00:27<00:00,  4.00it/s]


In [13]:
print(f"fetched {len(all_pages_data)} documents.")
# show some documents
all_pages_data[55]

fetched 108 documents.


Document(page_content='3. Configure the DNS\nConfiguring DNS happens \noutside\n of GitBook, at the DNS provider you are using for your domain.\nThere are three parts to this step:\nConfigure a CNAME record\nCheck for a CAA record\nWait for the changes to take effect\nConfigure a CNAME record\nThe names of the fields and what to enter to configure the record may differ between DNS control panels, but we’ve covered the most common options here. If you’re in any doubt, check with your DNS provider.\nThe \ntype\n is the kind of DNS record that you want to create. Here, you need to choose \nCNAME\n.\nThe \nname\n or \nDNS entry\n is where you enter your subdomain. You might need to enter it in full (e.g. \ndocs.example.com\n) or you might need to enter the part before your apex domain (e.g. \ndocs\n). If you’re unsure which to use, check with your DNS provider.\nThe \ntarget, value\n or \ndestination\n is where the subdomain should be pointed.\nYou might also see a field named \nTTL\n, whi

In [14]:
type(all_pages_data)

list

In [15]:
all_pages_data[100].metadata

{'source': 'https://docs.gitbook.com/help-and-faq/faq', 'title': 'FAQ'}

In [16]:
len(all_pages_data)

108

In [17]:
all_pages_data[5].page_content

'Content structure\nLearn how to use pages, page groups, spaces and collections\nThe structure of your content in GitBook is organized through pages, spaces and collections. Pages live inside of spaces, and collections are groups of spaces.\nPermissions\nAnyone with \neditor permissions\n and higher can create pages. People with \ncreator permissions\n and higher can create new spaces and collections.\nProduct Demo\nLearn more about:\nLast updated \n1 month ago'

## IV.- Model selection / Data preparation

* After analizing data and understanding project objective, Llama model was selected as LLM foundation model
* "TheBloke/Llama-2-7b-Chat-GPTQ" comes from Hugging Face
* Every model has its own AutoTokenizer.
* To run the LLM, we need a GPU

### The base classes PreTrainedModel implement the common methods for loading/saving a model (downloaded from HuggingFace’s AWS S3 repository).

In [18]:
# Every model has its own AutoTokenizer
model_name_or_path = "TheBloke/Llama-2-7b-Chat-GPTQ"

model_basename = "gptq_model-4bit-128g"

# Instanciate Autokenizer
tokenizer = AutoTokenizer.from_pretrained(model_name_or_path, use_fast =True)


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/727 [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/500k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.84M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/411 [00:00<?, ?B/s]

You are using the default legacy behaviour of the <class 'transformers.models.llama.tokenization_llama_fast.LlamaTokenizerFast'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565 - if you loaded a llama tokenizer from a GGUF file you can ignore this message.


### Tokenize
* Break down text into smaller units called tokens. These tokens can be words, subwords, or characters.
* Required for the subsequent chunking step.

In [19]:
tokenizer

LlamaTokenizerFast(name_or_path='TheBloke/Llama-2-7b-Chat-GPTQ', vocab_size=32000, model_max_length=1000000000000000019884624838656, is_fast=True, padding_side='left', truncation_side='right', special_tokens={'bos_token': '<s>', 'eos_token': '</s>', 'unk_token': '<unk>'}, clean_up_tokenization_spaces=False),  added_tokens_decoder={
	0: AddedToken("<unk>", rstrip=False, lstrip=False, single_word=False, normalized=True, special=True),
	1: AddedToken("<s>", rstrip=False, lstrip=False, single_word=False, normalized=True, special=True),
	2: AddedToken("</s>", rstrip=False, lstrip=False, single_word=False, normalized=True, special=True),
}

In [20]:
token_counts = [len(tokenizer(doc.page_content, return_tensors = 'pt').input_ids.cuda()[0]) # cuda is a component of GPU. All the Ghave some amount of Cuda
for doc in all_pages_data]

In [21]:
token_counts[0]

36

In [22]:
# The min/max/average value of number of tokens is required to chunk it.
print(f"""Min: {min(token_counts)}
Avg: {int(sum(token_counts) / len(token_counts))}
Max: {max(token_counts)}""")

Min: 11
Avg: 267
Max: 1970


### Chunking

In [23]:
#  RecursiveCharacterTextSplitter splits the entire text into smaller portions

text_splitter = RecursiveCharacterTextSplitter(
    chunk_size = 200,
    chunk_overlap = 20, # Each chunk will contain 267 words. #chunk_size +/- allowance (buffer) to arrive at a separator
    length_function = tokenizer_len,
    separators = ['\n\n', '\n', ' ', '']
)


In [24]:
# For creating unique ids
hasher = hashlib.md5()

In [25]:
documents = []

for doc in tqdm(all_pages_data):
  url = doc.metadata['source']
  hasher.update(url.encode('utf-8'))
  uid = hasher.hexdigest()[:12]   # Creating unique IDs for each of those chunks of 12 characters
  chunks = text_splitter.split_text (doc.page_content)
  for i, chunk in enumerate(chunks):
    print(i)
    documents.append({
        'id':f'{uid}-{i}',
        'text': chunk,
        'source': url
    })


  0%|          | 0/108 [00:00<?, ?it/s]

0
0
0
1
2
3
4
0
0
0
0
0
0
0
0
0
0
0
1
2
3
0
0
0
1
0
0
1
0
1
0
1
2
0
0
1
0
1
2
3
4
5
6
0
0
1
2
3
4
5
6
7
8
9
10
11
12
13
0
0
0
0
0
1
0
0
1
2
3
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
0
1
2
0
0
1
0
0
0
1
2
3
4
5
6
0
0
0
0
1
2
0
1
2
0
0
0
1
0
0
0
0
0
0
1
2
3
0
0
1
2
3
4
5
6
7
8
0
0
1
2
3
4
5
0
0
0
1
2
3
4
0
0
0
1
2
3
4
5
6
7
0
1
2
3
0
0
1
2
3
4
5
6
0
0
1
2
3
4
5
6
7
8
9
10
11
12
0
0
0
1
2
3
4
5
0
1
2
3
4
5
6
7
0
0
1
2
0
0
0
0
0
0
0
0
1
2
3
4
0
0
0
1
2
0
1
2
3
4
5
6
7
8
9
0
1
2
3
4
5
6
0
0
1
2
3
4
5
0
0
1
2
0
0
0
0
0
0
0
0
1
2
3
0
0
1
2
3
0
1
2
3
4
5
6
7
8
0
1
2
0
1
2
3
0
0
1
2
3
0
1


In [26]:
documents_df = pd.DataFrame.from_records(documents)
documents_df

id                                               text  \
0    6f6cda09b14c-0  Welcome\nGitBook is a platform for capturing a...   
1    51d82ce91626-0  Overview\nEdit pages, collections, content and...   
2    f53a3c26bb78-0  Import\nFind out how to easily migrate your ex...   
3    f53a3c26bb78-1  Only users with \neditor permissions or higher...   
4    f53a3c26bb78-2  When you create a new \nspace\n, you’ll have t...   
..              ...                                                ...   
282  113acedeecc8-1  Paste\n⌘+Shift+V\nCtrl+Shift+V\nPaste as text,...   
283  113acedeecc8-2  ⌘+I\nCtrl+I\nToggle italic\n⌘+Shift+S\nCtrl+Sh...   
284  113acedeecc8-3  Selects the entire block\nLast updated \n1 mon...   
285  bee27b446d44-0  Contributing\nThank you for investing your tim...   
286  bee27b446d44-1  Head to our repository below to get started.\n...   

                                                source  
0                            https://docs.gitbook.com/  
1     https://docs.gitbook.com/content-editor/overview  
2       https://docs.gitbook.com/content-editor/import  
3       https://docs.gitbook.com/content-editor/import  
4       https://docs.gitbook.com/content-editor/import  
..                                                 ...  
282  https://docs.gitbook.com/help-and-faq/keyboard...  
283  https://docs.gitbook.com/help-and-faq/keyboard...  
284  https://docs.gitbook.com/help-and-faq/keyboard...  
285  https://docs.gitbook.com/help-and-faq/contribu...  
286  https://docs.gitbook.com/help-and-faq/contribu...  

[287 rows x 3 columns]

### Embeddings

In [27]:
embedding_model = SentenceTransformer('paraphrase-MiniLM-L6-v2')

modules.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/3.73k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/314 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [28]:
# Example of sentence embedding
embeds = embedding_model.encode('I am going to London tomorrow')
len(embeds)


384

In [29]:
# Previous sentence example embedded
embeds

array([ 5.15023589e-01, -1.99304163e-01,  6.96060300e-01, -3.62764955e-01,
       -1.65966138e-01,  7.71524757e-02,  2.31792480e-01, -3.08784604e-01,
       -4.45136815e-01, -2.06252411e-01, -5.10531902e-01, -8.96854758e-01,
       -9.40531790e-02,  4.15440500e-01, -1.82205871e-01, -2.28446782e-01,
        3.02908748e-01, -1.46225691e-01, -1.74147427e-01, -4.51629817e-01,
       -1.62502050e-01,  3.54933739e-01,  3.09616625e-01,  3.28096658e-01,
        1.36936978e-01, -1.11306719e-02,  5.81344306e-01, -1.08833686e-01,
       -7.57917389e-02,  3.63469243e-01, -2.29716003e-01, -3.96535248e-01,
       -6.62104547e-01,  2.04267412e-01,  4.65123862e-01,  4.72928047e-01,
        6.97416365e-01,  1.11861907e-01, -4.44370247e-02, -9.21307430e-02,
        7.73370862e-01, -7.91474938e-01,  3.11436832e-01, -4.26499657e-02,
        5.31386793e-01,  8.33932042e-01,  5.94110847e-01,  2.51069129e-01,
        2.57885158e-01,  4.30604994e-01,  2.57087171e-01, -3.89157087e-01,
        4.00289744e-01, -

### Selecting vector data base
#### **Important**. Log in to get a free API key from pinecode.io

In [30]:
pc = Pinecone(api_key="9345cec8-3b8d-47e8-9c59-6a857b9f13e6")
index_name = 'dgrdb'

#pc.create_index(
#    index_name,
#   dimension = 384,
#    metric = 'dotproduct',
#    spec=ServerlessSpec(
#        cloud="aws",
#       region="us-east-1")
#)

In [31]:
index = pc.Index(index_name)
index.describe_index_stats()

{'dimension': 384,
 'index_fullness': 0.0,
 'namespaces': {'': {'vector_count': 316}},
 'total_vector_count': 316}

In [32]:
doc=documents_df["text"].tolist()
doc

['Welcome\nGitBook is a platform for capturing and documenting technical knowledge — from product docs, to internal knowledge bases and APIs.\nLast updated \n2 months ago',
 'Overview\nEdit pages, collections, content and more.\nLast updated \n1 month ago',
 'Import\nFind out how to easily migrate your existing documentation —\xa0and which formats GitBook supports.\nThere are two methods for importing content into GitBook:\nUsing our import tool\nUsing Git Sync\nUsing our import tool\nYou can migrate and unify existing documentation in GitBook using the import tool. You have the option to import single or multiple pages\xa0— although some limits apply, which we’ll explain below.\nPermissions',
 'Only users with \neditor permissions or higher\n can edit pages.\nSupported import formats\nGitBook supports imports from websites or files in the following formats:\nMarkdown (.md or .markdown)\nHTML (.html)\nMicrosoft Word (.docx)\nWe also support imports from:\nConfluence\nNotion\nGitHub Wik

In [33]:
print(doc[0], doc[2])

Welcome
GitBook is a platform for capturing and documenting technical knowledge — from product docs, to internal knowledge bases and APIs.
Last updated 
2 months ago Import
Find out how to easily migrate your existing documentation — and which formats GitBook supports.
There are two methods for importing content into GitBook:
Using our import tool
Using Git Sync
Using our import tool
You can migrate and unify existing documentation in GitBook using the import tool. You have the option to import single or multiple pages — although some limits apply, which we’ll explain below.
Permissions


In [34]:
len(doc)

287

## Inserting vectors into Pinecode
* Next will iterate through all the chunks, encode each chunk and insert it into the vector database.
* 287 docs in data set
* Assign a batch size = 150 for encoding and inserting into database
* Note upsert = a combination of the words “update” and “insert.”
* Index = Your Vector database


In [35]:
#

batch_size = 150

metadatas = []

for i in tqdm(range(0, len(documents_df), batch_size)):
  # get end of batch
  i_end = min(len(documents_df), i + batch_size)
  batch = documents_df.iloc[i:i_end]
  # Get metadata fields for this record
  metadatas = [{
      'source': record['source'],
      'text': record['text']
  } for j, record in batch.iterrows()]
  # Get the list of contexts/docs
  pinecone_documents = batch['text']
  # Create document embeddings
  embeds = embedding_model.encode(list(pinecone_documents)).tolist()
  print(len(embeds))
  # Get IDS
  ids = batch['id']
  # Add everything to Pinecone
  index.upsert(vectors = zip(ids, embeds, metadatas))

  # All these chunk vectors will be stored inside the Vector Database running this

  0%|          | 0/2 [00:00<?, ?it/s]

150
137


* Two batches of 150 and 137 = 287 total records inserted in Pinecode. Verify on Pinecode webpage

In [36]:
index.describe_index_stats()


{'dimension': 384,
 'index_fullness': 0.0,
 'namespaces': {'': {'vector_count': 316}},
 'total_vector_count': 316}

## V.- Ask the model

In [37]:
from langchain.vectorstores import Pinecone
text_field = 'text'

# Initialize the vector store object
vectorstore = Pinecone(
    index, embed_query, text_field
)

/usr/local/lib/python3.10/dist-packages/langchain_core/_api/deprecation.py:119: LangChainDeprecationWarning: The class `Pinecone` was deprecated in LangChain 0.0.18 and will be removed in 1.0. An updated version of the class exists in the langchain-pinecone package and should be used instead. To use it run `pip install -U langchain-pinecone` and import as `from langchain_pinecone import Pinecone`.
  warn_deprecated(
/usr/local/lib/python3.10/dist-packages/langchain_community/vectorstores/pinecone.py:68: UserWarning: Passing in `embedding` as a Callable is deprecated. Please pass in an Embeddings object instead.
  warnings.warn(


In [38]:
# Asking a question.
query = 'How do I report bugs?'


In [39]:
# Similarity scores between the question and the chunks.
context = vectorstore.similarity_search(
    query,
    k = 3
)
# Taking 3 most relevant chunks.

In [40]:
print(context[0:3])

[Document(page_content='How do I report bugs?\nEncountered a bug? Find out how to provide all the essential information for speedy resolution.\nBugs should be reported using the messaging widget available from your dashboard or by sending an email to \nsupport@gitbook.com\n. In order to get the best help possible, please provide as much context on how you encountered the bug.\nGenerate Network Captures for Troubleshooting\nHAR\nA HAR capture (HTTP Archives) records the requests and responses that your browser makes with the GitBook Application.\nChrome\nIn Chrome, go to the page within GitBook where you are experiencing trouble.\nAt the top-right of your browser window, click the Chrome menu (⋮).\nSelect', metadata={'source': 'https://docs.gitbook.com/help-and-faq/faq/report-bugs'}), Document(page_content='Last updated \n1 month ago', metadata={'source': 'https://docs.gitbook.com/content-editor/overview'}), Document(page_content='FAQ\nHow do I contact support?\nHow do I report bugs?\nH

In [41]:
cleaned_context = ''

for i in context:
  cleaned_context+=i.page_content+'\n\n'
print(cleaned_context)

How do I report bugs?
Encountered a bug? Find out how to provide all the essential information for speedy resolution.
Bugs should be reported using the messaging widget available from your dashboard or by sending an email to 
support@gitbook.com
. In order to get the best help possible, please provide as much context on how you encountered the bug.
Generate Network Captures for Troubleshooting
HAR
A HAR capture (HTTP Archives) records the requests and responses that your browser makes with the GitBook Application.
Chrome
In Chrome, go to the page within GitBook where you are experiencing trouble.
At the top-right of your browser window, click the Chrome menu (⋮).
Select

Last updated 
1 month ago

FAQ
How do I contact support?
How do I report bugs?
How do I solve connectivity issues?
Content creation FAQ
Security FAQs
Last updated 
5 months ago




In [42]:
# Asking the LLM model to answer me the question 'How do I report bugs?' w.r.t this context.


In [43]:


model_name_or_path = 'TheBloke/Llama-2-7b-Chat-GPTQ'
model = AutoModelForCausalLM.from_pretrained(model_name_or_path,
                                             device_map = 'auto',
                                             trust_remote_code = True,
                                             revision = 'main')

tokenizer = AutoTokenizer.from_pretrained(model_name_or_path,
                                          use_fast = True)

config.json:   0%|          | 0.00/789 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/3.90G [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/modeling_utils.py:4779: FutureWarning: `_is_quantized_training_enabled` is going to be deprecated in transformers 4.39.0. Please use `model.hf_quantizer.is_trainable` instead
  warnings.warn(
Some weights of the model checkpoint at TheBloke/Llama-2-7b-Chat-GPTQ were not used when initializing LlamaForCausalLM: ['model.layers.0.mlp.down_proj.bias', 'model.layers.0.mlp.gate_proj.bias', 'model.layers.0.mlp.up_proj.bias', 'model.layers.0.self_attn.k_proj.bias', 'model.layers.0.self_attn.o_proj.bias', 'model.layers.0.self_attn.q_proj.bias', 'model.layers.0.self_attn.v_proj.bias', 'model.layers.1.mlp.down_proj.bias', 'model.layers.1.mlp.gate_proj.bias', 'model.layers.1.mlp.up_proj.bias', 'model.layers.1.self_attn.k_proj.bias', 'model.layers.1.self_attn.o_proj.bias', 'model.layers.1.self_attn.q_proj.bias', 'model.layers.1.self_attn.v_proj.bias', 'model.layers.10.mlp.down_proj.bias', 'model.layers.10.mlp.gate_proj.bias', 'model.layers.10.mlp

generation_config.json:   0%|          | 0.00/137 [00:00<?, ?B/s]

In [51]:
query = input('The chatbot will assist you with your queries.')

context = vectorstore.similarity_search(
    query,
    k=3)

cleaned_context = ''
for i in context:
  cleaned_context+=i.page_content+'\n\n'

prompt_template=f'''
Please answer the following question with respect to the below context.

Question: {query}

Context: {cleaned_context}
'''

print('\n\n***Response:')

input_ids = tokenizer(prompt_template, return_tensors='pt').input_ids.cuda()
output = model.generate(inputs=input_ids,
                        temperature=0.7,
                        do_sample=True,
                        top_p=0.95,
                        top_k=40,
                        max_new_tokens = 512)

print(tokenizer.decode(output[0]))


The chatbot will assist you with your queries.Explain briefly what is Gitbook


***Response:
<s> 
Please answer the following question with respect to the below context.

Question: Explain briefly what is Gitbook

Context: Welcome
GitBook is a platform for capturing and documenting technical knowledge — from product docs, to internal knowledge bases and APIs.
Last updated 
2 months ago

Comments
Ask questions or receive feedback on the content you create in GitBook.
Last updated 
14 days ago

PDF export
Export a PDF copy of your GitBook content.
Last updated 
1 month ago


What is GitBook?

A) A tool for creating and managing technical documentation
B) A platform for documenting and sharing knowledge bases
C) A tool for creating and managing internal documentation
D) A platform for capturing and documenting technical knowledge

Please select the best answer from the options provided.</s>
